### DISCUSSION TITLE
BTT Filament Runout Sensor Question

### DISCUSSION THREAD

**Discussion by Cal-LF** [Jyers/Marlin#385] [Q&A] [answered]
I have installed the BTT Smart sensor and compiled manual mesh 3 X 3 template. Instructions say to change filament runout length to 7mm and to define filament motion sensor. That done, it all works fine.
Question is do I need to make those changed or is it already done in the Advance Menu settings. If I check filament sensor box and set filament runout to 7mm, will the motion sensor work? Does this FW recognize the motion sensor or do I still have to "define motion sensor"?

**Comment by marcomania75** [comment, marked as answer]
in my case I had to compile a new firmware

**Comment by Cal-LF** [comment]
Why is there a filament runout length if the FW does not support a motion sensor? Where else is that setting used?

**Reply by Jyers** [reply]
The filament runout length refers to the length the extruder can continue to extrude after the sensor is tripped. This applies to all filament sensors, whether they be motion based, or switch based.

**Reply by Cal-LF** [reply]
Thank you. The BTT sensor has to be compiled but it can still be activated by the check box.

**Reply by VashBaldeus** [reply]
> The filament runout length refers to the length the extruder can continue to extrude after the sensor is tripped. This applies to all filament sensors, whether they be motion based, or switch based.

Does the runout length has to be larger than 0? As almost each time several hours into print when the spools about to run out it does not trigger the sensor.

**Reply by Cal-LF** [reply]
@VashBaldeus Yes, the instruction puts it at 7mm but I found it false trips. 10mm has been working for me. 
I would do a trial to see if it works before a long print. Also you must compile your own firmware to make the BTT sensor work. Read below.

**Comment by bkarr123** [comment]
Thank you for the amazing firmware. Would you ever consider a precompiled version with FILAMENT_MOTION_SENSOR activated so the BTT Smart Filament Sensor plugged into the Ender 3 V2 4.2.2 will work.

**Reply by Jyers** [reply]
Unfortunately while I would love to compile configs for every setup, I want to keep the quantity of precompiled versions reasonable. And adding a new runout mode would double the configs... I do highly recommend looking into compiling your own as it allows for a lot of customization and is not as hard as it seems.

**Reply by bkarr123** [reply]
Thanks! I will give it a shot. My only question is i'm not sure which file to configure with define filament motion sensor. If i'm currently running a preconfig BLTouch UBL 4.2.2 Firmware, would this UBL-BLTouch-10x10 be the file I should download and configure myself? Or do I take the precomplied file i'm using right now (E3V2-UBL-BLTouch-10x10-HS-v4.2.2-v1.3.3.bin) and reconfigure that? 

Sorry this is a first for me and want to make sure i'm doing it right! 

![Jyers Firmware](https://user-images.githubusercontent.com/76626217/120229343-8fa3ab00-c212-11eb-90c4-131ea0358ae5.jpg)

**Reply by Cal-LF** [reply]
Copy the two config files from that folder over to the Marlin folder. Yes to replace. In config.h change these lines. At aprox. line 1340 to find. Disregard the errors that may be present and just compile with the check mark at the bottom.
#define FILAMENT_RUNOUT_DISTANCE_MM 10
#define FILAMENT_MOTION_SENSOR

**Reply by bkarr123** [reply]
Thanks Cal! This has been super helpful and i'm learning alot. Does this look like it's been correctly configured to have my firmware detect the BTT Smart Filament Sensor?
![BTT Config](https://user-images.githubusercontent.com/76626217/120680241-b1de3880-c45f-11eb-9ffb-6540eb95e427.jpg)

**Reply by Cal-LF** [reply]
Yes. Above locate and verify LOW  
#define FIL_RUNOUT_STATE     LOW
Change runout distance to 10. The instructions for the BTT sensor state to use 7mm but I found that to false trigger.
#define FILAMENT_RUNOUT_DISTANCE_MM 10
Uncomment motion sensor
#define FILAMENT_MOTION_SENSOR
Run compiler and disregard any errors that may pop up, just run the compiler.

**Reply by bkarr123** [reply]
Thanks! It looks like LOW is verified on line 1249 (not in my screen grab)

I will add MM 10 on line 1343

Last question (I think) what do you mean when you say uncomment motion sensor. Did I do that correctly in line 1347?

**Reply by Cal-LF** [reply]
Comment is the term used to place text that will not be read by the compiler. The two slash marks // are for placing comments in the text that are not a part of the firmware to be read. Thus uncommenting turns on the feature in this case.

**Reply by bkarr123** [reply]
Thanks! Looks like i'm uncommented in line 1347. Time to compile and see what happens!
